### ISSUE TITLE
Compilation fails with latest llvm (after revision 368707)

### ISSUE THREAD

**Issue by ToucheSir** [tensorflow/mlir#81]
It seems https://github.com/llvm/llvm-project/commit/52a34a78d9aff1bb5e66e7c32490229ea177e075 (made yesterday, what timing!) breaks `lib/ExecutionEngine/ExecutionEngine.cpp`:

```
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp: In constructor ‘mlir::impl::SearchGenerator::SearchGenerator(char)’:
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp:87:31: error: no matching function for call to ‘llvm::orc::DynamicLibrarySearchGenerator::DynamicLibrarySearchGenerator(std::unique_ptr<llvm::orc::DynamicLibrarySearchGenerator>)’
                 GlobalPrefix))) {}
                               ^
In file included from .../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp:28:0:
.../llvm-project/llvm/include/llvm/ExecutionEngine/Orc/ExecutionUtils.h:253:3: note: candidate: llvm::orc::DynamicLibrarySearchGenerator::DynamicLibrarySearchGenerator(llvm::sys::DynamicLibrary, char, llvm::orc::DynamicLibrarySearchGenerator::SymbolPredicate)
   DynamicLibrarySearchGenerator(sys::DynamicLibrary Dylib, char GlobalPrefix,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.../llvm-project/llvm/include/llvm/ExecutionEngine/Orc/ExecutionUtils.h:253:3: note:   candidate expects 3 arguments, 1 provided
.../llvm-project/llvm/include/llvm/ExecutionEngine/Orc/ExecutionUtils.h:243:7: note: candidate: llvm::orc::DynamicLibrarySearchGenerator::DynamicLibrarySearchGenerator(const llvm::orc::DynamicLibrarySearchGenerator&)
 class DynamicLibrarySearchGenerator : public JITDylib::DefinitionGenerator {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.../llvm-project/llvm/include/llvm/ExecutionEngine/Orc/ExecutionUtils.h:243:7: note:   no known conversion for argument 1 from ‘std::unique_ptr<llvm::orc::DynamicLibrarySearchGenerator>’ to ‘const llvm::orc::DynamicLibrarySearchGenerator&’
.../llvm-project/llvm/include/llvm/ExecutionEngine/Orc/ExecutionUtils.h:243:7: note: candidate: llvm::orc::DynamicLibrarySearchGenerator::DynamicLibrarySearchGenerator(llvm::orc::DynamicLibrarySearchGenerator&&)
.../llvm-project/llvm/include/llvm/ExecutionEngine/Orc/ExecutionUtils.h:243:7: note:   no known conversion for argument 1 from ‘std::unique_ptr<llvm::orc::DynamicLibrarySearchGenerator>’ to ‘llvm::orc::DynamicLibrarySearchGenerator&&’
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp: In member function ‘llvm::Expected<llvm::DenseSet<llvm::orc::SymbolStringPtr> > mlir::impl::SearchGenerator::operator()(llvm::orc::JITDylib&, const SymbolNameSet&)’:
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp:94:42: error: no match for call to ‘(llvm::orc::DynamicLibrarySearchGenerator) (llvm::orc::JITDylib&, const SymbolNameSet&)’
     auto res = defaultGenerator(JD, Names);
                                          ^
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp: In constructor ‘mlir::impl::OrcJIT::OrcJIT(llvm::orc::JITTargetMachineBuilder, llvm::DataLayout, mlir::impl::OrcJIT::IRTransformer, llvm::ArrayRef<llvm::StringRef>)’:
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp:142:31: error: ‘class llvm::orc::JITDylib’ has no member named ‘setGenerator’; did you mean ‘addGenerator’?
     session.getMainJITDylib().setGenerator(
                               ^~~~~~~~~~~~
                               addGenerator
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp: In member function ‘void mlir::impl::OrcJIT::loadLibraries(llvm::ArrayRef<llvm::StringRef>)’:
.../llvm-project/llvm/projects/mlir/lib/ExecutionEngine/ExecutionEngine.cpp:226:8: error: ‘class llvm::orc::JITDylib’ has no member named ‘setGenerator’; did you mean ‘addGenerator’?
     JD.setGenerator(loaded.get());
        ^~~~~~~~~~~~
        addGenerator

```

In short:
1. `llvm::orc::JITDylib.setGenerator` seems to have been replaced by `.addGenerator`
2. `SearchGenerator` now has to extend `llvm::orc::JITDylib::DefinitionGenerator`
3. Many of the generator-related APIs now take and return `unique_ptr`s

**Comment by ftynse** [comment]
This has been fixed and will be visible after our systems integrate an LLVM version that is recent enough, thanks.

**Comment by ftynse** [comment]
This specific issue has been fixed by 5dd2ac4016d2bcd557fc98cea89ecb6915452f0c.  CI build is still broken because of the compiler not having C++14 support.

**Comment by ToucheSir** [comment]
Thanks for the quick fix!

As someone who is still wrapping their head around the codebase, just a question about https://github.com/tensorflow/mlir/commit/5dd2ac4016d2bcd557fc98cea89ecb6915452f0c#diff-c9108781315ffa6fb84366eab36cd6e4R143: Is the mlir `SearchGenerator` wrapper class used outside of `OrcJIT`? One of the issues I ran into trying to resolve the compilation errors locally was that `JITDylib.addGenerator` now expects a subclass of `JITDylib::DefinitionGenerator`, which seems to suggest that SearchGenerator is just meant as a decorator around another `DefinitionGenerator`?
